src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -234,26 +158,179 @@ void addArg(RefKind refKind, BoundExpression expression)
                             HashSet<DiagnosticInfo> useSiteDiagnostics = null;
                             Conversion conversion = _factory.Compilation.Conversions.ClassifyBuiltInConversion(inputType, output.Type, ref useSiteDiagnostics);
                             _localRewriter._diagnostics.Add(t.Syntax, useSiteDiagnostics);
+                            BoundExpression evaluated;
                             if (conversion.Exists)
                             {
-                                _sideEffectBuilder.Add(_factory.AssignmentExpression(output, _factory.Convert(type, input, conversion)));
+                                if (conversion.Kind == ConversionKind.ExplicitNullable &&
+                                    inputType.GetNullableUnderlyingType().Equals(output.Type, TypeCompareKind.AllIgnoreOptions) &&
+                                    _localRewriter.TryGetNullableMethod(t.Syntax, inputType, SpecialMember.System_Nullable_T_GetValueOrDefault, out MethodSymbol getValueOrDefault))
+                                {
+                                    // As a special case, since the null test has already been done we can use Nullable<T>.GetValueOrDefault
+                                    evaluated = _factory.Call(input, getValueOrDefault);
+                                }
+                                else
+                                {
+                                    evaluated = _factory.Convert(type, input, conversion);
+                                }
                             }
                             else
                             {
-                                _sideEffectBuilder.Add(_factory.AssignmentExpression(output, _factory.As(input, type)));
+                                evaluated = _factory.As(input, type);
                             }
+
+                            return _factory.AssignmentExpression(output, evaluated);
                         }
-                        break;
+
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(evaluation);
                 }
+            }
 
-                return false;
+            /// <summary>
+            /// Return the boolean test required for the given decision. Returns `null` if the decision is trivially true.
+            /// </summary>
+            protected BoundExpression LowerDecision(BoundDagDecision decision)
+            {
+                BoundExpression input = _tempAllocator.GetTemp(decision.Input);
+                switch (decision)
+                {
+                    case BoundNonNullDecision d:
+                        // If the actual input is a constant, short-circuit this test
+                        if (d.Input == _inputTemp && _loweredInput.ConstantValue != null)
```

Comment:
It feels like it would have been nice to consider this earlier. Is that not possible because we haven't done constant folding until now? Or is it just that the peephole optimization is simpler to do in lowering?

Reply:
I believe we have done it earlier, and that this code has no effect. If I can verify that fact, I will assert it. Recorded in [https://github.com/dotnet/roslyn/issues/25663](https://github.com/dotnet/roslyn/issues/25663)

---
In reply to: [176228619](https://github.com/dotnet/roslyn/pull/24059#discussion_r176228619) [](ancestors = 176228619)